### PR TITLE
fix(repository-json-schema): allows string-based order filter

### PR DESCRIPTION
**Pull request by achrinza**
fixes: #6140

Signed-off-by: Rifa Achrinza <25147899+achrinza@users.noreply.github.com>

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- API Documentation in code was updated
- Documentation in [/docs/site](../tree/master/docs/site) was updated
- Affected artifact templates in `packages/cli` were updated
- Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Review by hacksparrow — APPROVED** [review]

**Comment by hacksparrow** [comment]
Please let @jannyHou and @agnes512 take a look too.

File: packages/repository-json-schema/src/__tests__/unit/filter-json-schema.unit.ts [hunk 1]
```diff
@@ -172,15 +180,26 @@ describe('getFilterJsonSchemaFor', () => {
...
         keyword: 'type',
         dataPath: '.order',
         message: 'should be array',
       },
+      {
+        keyword: 'oneOf',
+        dataPath: '.order',
+        params: {passingSchemas: null},
+        message: 'should match exactly one schema in oneOf',
+      },
```
Comment (bajtos) [inline packages/repository-json-schema/src/__tests__/unit/filter-json-schema.unit.ts:202]: Hmm, if I am reading this correctly, the error message is not very descriptive. As I understand this data, it says:

```
.order should be string
.order should be array
.order should match exactly on schema in oneOf
```

I find that a bit confusing, but I guess it's a problem in AJV, not something we can fix at LoopBack side. Would you mind to open an issue in [AJV bug tracker](https://github.com/ajv-validator/ajv/issues)? I did a quick search and did not find any relevant issue.

**Review by bajtos — COMMENTED** [review]
Nice UX improvement!

Can you please add a test that will send a request using the new "order" style and verify that this style is actually supported by `@loopback/repository`? Do we need to update TypeScript definition for `Filter.order` property too?

**Comment by achrinza** [comment]
> Do we need to update TypeScript definition for Filter.order property too?

Nope, it already accepts a single string (among others), and `FilterBuilder` always builds an array and already has the tests in place.

> Can you please add a test that will send a request using the new "order" style and verify that this style is actually supported by @loopback/repository?

Not sure where to add that test, any recommendations?

**Comment by bajtos** [comment]
> > Do we need to update TypeScript definition for Filter.order property too?
> 
> Nope, it already accepts a single string (among others), and `FilterBuilder` always builds an array and already has the tests in place.

Thanks for verifying!

> > Can you please add a test that will send a request using the new "order" style and verify that this style is actually supported by @loopback/repository?
> 
> Not sure where to add that test, any recommendations?

Hmm, that's a good question. I don't see any existing test that would invoke REST API with `filter.order` set. 

I think our current approach is to put end-to-end tests involving REST + Repository layers into `examples/todo` test case. Here is an existing test to verify `filter` support:

https://github.com/strongloop/loopback-next/blob/9b2a0c93a41f78e1d810d42cc48ad93f3d988480/examples/todo/src/__tests__/acceptance/todo.acceptance.ts#L200-L212

Here are commits that are already using `examples/todo` for REST+Repository integration tests: 7e1acfc1e72, a4ef64037

I am proposing to add a new test to `todo.acceptance.ts`. In the test, you can create multiple Todo item instances, and then make a query to get them ordered using given criteria. I think it would best to write more than one test and verify different ways how the filter can be encoded - at least `filter[order]=title+DESC` and `filter={"order":"title+DESC"}`.

File: examples/todo/src/__tests__/acceptance/todo.acceptance.ts [hunk 2]
```diff
@@ -222,6 +222,110 @@ describe('TodoApplication', () => {
...
+    });
+
+    await client
+      .get('/todos')
+      .query({filter: {order: 'title DESC'}})
+      .expect(200, [
+        toJSON(todoCompleted),
+        toJSON(todoCompleted2),
+        toJSON(todoInProgress),
+      ]);
```
Comment (bajtos) [inline examples/todo/src/__tests__/acceptance/todo.acceptance.ts:248]: Nitpick - feel free to ignore.

The code can be simplified as follows.

```suggestion
      .expect(200, toJSON([
        todoCompleted,
        todoCompleted2,
        todoInProgress,
      ]));
```

Same comment applies to other new tests too.

**Review by bajtos — APPROVED** [review]
Excellent!

I have one nitpick to consider or ignore, the patch LGTM.

**Comment by bajtos** [comment]
(Please clean up the commit history before landing.)

**Pull request merged**